MSC4383: Client-Server Discovery of Server Version

jevolk opened the pull request:
[rendered](https://github.com/matrix-construct/matrix-refreshed/blob/jevolk/client-server-info/proposals/4383-client-server-info.md)

## Implementations

- **Server**: [Tuwunel](https://github.com/matrix-construct/tuwunel) (shipped in v1.6.2), populating the `server` object on `GET /_matrix/client/versions` and advertising `net.zemos.msc4383` in `unstable_features`.
- **Protocol types**: [ruma/ruma#2495](https://github.com/ruma/ruma/pull/2495) adds the `Server` struct and the `Response.server` field behind the `unstable-msc4383` Cargo feature.
- **Client SDK**: [matrix-org/matrix-rust-sdk#6622](https://github.com/matrix-org/matrix-rust-sdk/pull/6622) rewrites `Client::server_vendor_info` to prefer the client-server source, falling back to `GET /_matrix/federation/v1/version` only when the object is absent. The feature is enabled by default in `matrix-sdk-ffi`, so applications consuming the FFI binding acquire the client-server path on rebuild without an app-side flag.

## Pre-acceptance checklist

- [x] Are appropriate implementation(s) specified in the MSC's PR description? (See above.)
- [x] Are all MSCs that this MSC depends on already accepted? (No dependencies.)
- [x] For each endpoint that is introduced or modified:
  - [x] Have authentication requirements been specified? (`GET /_matrix/client/versions` remains unauthenticated.)
  - [x] Have rate-limiting requirements been specified? (Not rate-limited.)
  - [x] Have guest access requirements been specified? (Guest-access restrictions are not applicable.)
  - [x] Are error responses specified? (No new error responses; the field is purely additive on the existing response body.)
    - [x] Does each error case have a specified `errcode` and HTTP status code? (N/A.)
      - [x] If a new `errcode` is introduced, is it clear that it is new? (N/A.)
- [x] Will the MSC require a new room version, and if so, has that been made clear? (No new room version.)
  - [x] Is the reason for a new room version clearly stated? (N/A.)
- [x] Are backwards-compatibility concerns appropriately addressed? (See §Backwards Compatibility: additive wire shape, `unstable_features` advertise gate, and SDK migration path preserving the existing `Client::server_vendor_info` contract.)
- [x] Are the endpoint conventions honoured?
  - [x] Do HTTP endpoints `use_underscores_like_this`? (Existing endpoint; convention preserved.)
  - [x] Will the endpoint return unbounded data? If so, has pagination been considered? (Fixed-shape object; pagination not applicable.)
  - [x] If the endpoint utilises pagination, is it consistent with the appendices? (N/A.)
- [x] An introduction exists and clearly outlines the problem being solved. (§Introduction.)
- [ ] All outstanding threads are resolved.
  - [x] All feedback is incorporated into the proposal text itself, either as a fix or noted as an alternative.
- [x] While the exact sections do not need to be present, the details implied by the proposal template are covered:
  - [x] Introduction
  - [x] Proposal text
  - [x] Potential issues
  - [x] Alternatives
  - [x] Dependencies (none; omitted)
- [x] Stable identifiers are used throughout the proposal, except for the unstable prefix section.
  - [x] Unstable prefixes consider the awkward accepted-but-not-merged state (`net.zemos.msc4383.stable` companion flag listed in §Unstable prefix).
  - [x] Chosen unstable prefixes do not pollute any global namespace (vendor prefix is the controlled reverse-DNS `net.zemos.*`).
- [x] Changes have applicable Sign Off from all authors/editors/contributors. Per-commit Signed-off-by on every new commit.
- [x] There is a dedicated "Security Considerations" section.

jevolk committed:
MSC4383: Client-Server Discovery of Server Version

turt2live commented on proposals/4383-client-server-info.md:
Implementation requirements:

* [ ] Client (ideally multiple)
* [ ] Server

turt2live commented on proposals/4383-client-server-info.md:
```diff
@@ -0,0 +1,80 @@
+# MSC4383: Client-Server Discovery of Server Version
+
```
This proposal appears to be lacking a problem statement and/or introduction, making it difficult to review with an aim to include it in the spec.

jevolk replied:
The problem is stated in the [Alternatives](https://github.com/matrix-construct/matrix-refreshed/blob/jevolk/client-server-info/proposals/4383-client-server-info.md#alternatives) section. I didn't feel the need to make a redundant introduction given the status quo alternative is prominent and it's a small proposal otherwise. Should the first paragraph be moved to the top?

turt2live replied:
Relevant information and background should always be first. I didn't feel the paragraph in the Alternatives was suited as an introduction because it lacks background and context.

jevolk commented on proposals/4383-client-server-info.md:
```diff
@@ -0,0 +1,92 @@
+Client applications utilizing the Matrix SDK[^6], a Client-Server[^7] library, have been observed
+making requests to `GET /_matrix/federation/v1/version`[^2]. It is the only known example of a
+cross-interface request to have any possible[^5] use to any implementation[^3], motivating this
+proposal to reestablish a properly clean partition between client and federation interfaces.
```
In the found case, the information is logged and not used to determine application-logic. The logs are then included in rageshakes.

The query is only to the local server, as remote servers would require server name resolution which the client is not capable of—arguably the local server also requires the resolution for the federation interface which is itself another concern for why the status quo is untenable.

Johennes replied:
Sorry, no. I just meant to use that as an example to illustrate that without describing the use case it's not clear why the proposed solution is the best one.

jevolk committed:
Move description of use case from footnotes to introduction.

jevolk commented:
Implemented in Tuwunel (shipped in v1.6.2). `GET /_matrix/client/versions` includes the unstable `net.zemos.msc4383.server` object populated with the homeserver name and version string, alongside the `net.zemos.msc4383` flag in `unstable_features`.

Source: [src/api/client/versions.rs#L32-L38](https://github.com/matrix-construct/tuwunel/blob/51c19f3ffd21d2e02d25ce11ef9b035f0372d01d/src/api/client/versions.rs#L32-L38).

jevolk committed:
chore: sign-off, heading rename, JSON tags, footnote wrap.

Folds several housekeeping items into a single chore:

- Retroactively adds Signed-off-by trailers for prior commits
  63cdb4b1 and 7904077d, rather than amending and force-pushing
  the original commits.
- Renames the section heading "Security Concerns" to
  "Security Considerations" to match the upstream MSC template.
- Switches the two fenced example blocks in the Proposal section
  from bare ``` to ```json so renderers highlight the JSON shapes.
- Splits the long matrix-rust-sdk source footnote ([^3]) URL onto
  its own line under the footnote marker. The URL itself is 137
  characters and cannot be wrapped without breaking the link;
  the split keeps prose lines clear and lets the URL sit as a
  delimited footnote block.

Signed-off-by: Jason Volk <jason@zemos.net>
Pre-trailer-sha: 63cdb4b1b4ed1f0e4ee0ef8acc1eb7c3e7e72145
Pre-trailer-sha: 7904077d78e90d85f6ce56c0937bca2f5ffb92e4

jevolk committed:
Cite Tuwunel, Ruma, and matrix-rust-sdk as implementations.

Adds an Implementations section listing the shipped Tuwunel server
support, the open Ruma protocol-types pull request, and the open
matrix-rust-sdk client pull request. Footnote anchors carry the
linking URLs.

Signed-off-by: Jason Volk <jason@zemos.net>

jevolk committed:
Specify endpoint conventions for /_matrix/client/versions.

States that GET /_matrix/client/versions remains unauthenticated, is
not rate-limited, that guest-access restrictions are not applicable,
and that no new error responses are introduced. Closes the
endpoint-conventions clauses of the SCT checklist for this MSC.

Signed-off-by: Jason Volk <jason@zemos.net>

jevolk committed:
Address backwards compatibility.

Adds a Backwards Compatibility section covering the additive wire
shape, the net.zemos.msc4383 detection flag, and the SDK fallback
pattern that lets application consumers acquire the partition
correction with no code change. Closes the backwards-compatibility
clause of the SCT checklist for this MSC.

Signed-off-by: Jason Volk <jason@zemos.net>

jevolk committed:
Rework introduction; adopt British spelling.

Adds a non-technical lead paragraph addressing turt2live's review
request that background and context appear first. Folds Johennes's
review clarifications into a dedicated paragraph: the use case is
logging only (the identification is included in rageshakes and
similar diagnostic submissions, not consulted to determine
application behaviour) and the query targets the user's own
homeserver only. Renames the section heading from "Motivation" to
"Introduction" per upstream template convention.

Incidental fixes carried by the rewrite: corrects the "reestablises"
typo to "reestablishes", drops the incorrect "as" from "rendering
the status quo as unreliable", removes the hyphen from
"site-administrators", and switches "utilizing" to "utilising" so
the document is consistently British spelling.

Signed-off-by: Jason Volk <jason@zemos.net>

jevolk committed:
Cite element-web's rageshake workaround chain.

Adds a parallel citation to element-web's rageshake collector,
which calls the Synapse-specific administrative endpoint
GET /_synapse/admin/v1/server_version with a federation-side
/version fallback. The pair illustrates the two workaround
shapes in production client code: cross-interface (matrix-rust-sdk)
and vendor lock-in (element-web), neither of which functions when
its corresponding interface is restricted at the deployment edge.

Signed-off-by: Jason Volk <jason@zemos.net>

jevolk committed:
Add RFC 2119 binding requirements.

Adds explicit MUST/SHOULD/MAY/MUST NOT obligations covering server
population of the new server object, parity with the federation
/version response (with an escape valve for federation-disabled
deployments), and the prohibition on clients gating application
logic on the object's contents.

Signed-off-by: Jason Volk <jason@zemos.net>

jevolk committed:
Add stable companion flag to unstable prefix.

Extends the unstable-prefix table with rows for the
net.zemos.msc4383 support advertisement and the
net.zemos.msc4383.stable companion advertisement that signals the
accepted-but-not-merged interval, so clients can detect support and
migrate reads to the unprefixed server key without coordination.

Signed-off-by: Jason Volk <jason@zemos.net>

jevolk committed:
Apply pre-submission review corrections.

Folds minor textual corrections identified during a pre-submission review.

Signed-off-by: Jason Volk <jason@zemos.net>
